cnd/src/btsieve/mod.rs

```diff
@@ -42,3 +46,126 @@ pub trait ReceiptByHash: Send + Sync + 'static {
 pub trait Predates {
     fn predates(&self, timestamp: NaiveDateTime) -> bool;
 }
+
+pub trait BlockHashes<H> {
+    fn block_hash(&self) -> H;
+    fn prev_block_hash(&self) -> H;
+}
+
+/// This function uses the `connector` to find blocks relevant to a swap.  To do
+/// this we must get the latest block, for each latest block we receive we must
+/// ensure that we saw its parent i.e., that we did not miss any blocks between
+/// this latest block and the previous latest block we received.  Finally, we
+/// must also get each block back until the time that the swap started i.e.,
+/// look into the past (in case any action occurred on chain while we were not
+/// watching).
+///
+/// It yields those blocks as part of the process.
+pub async fn find_relevant_blocks<C, B, H>(
+    mut connector: C,
+    co: &Co<B>,
+    start_of_swap: NaiveDateTime,
+) -> anyhow::Result<Never>
+where
+    C: LatestBlock<Block = B> + BlockByHash<Block = B, BlockHash = H> + Clone,
+    B: Predates + BlockHashes<H> + Clone,
+    H: Eq + std::hash::Hash + Copy,
```

Comment:
Within this scope, you should be able to remove that.
```suggestion
    H: Eq + Hash + Copy,
```